checks.d/hdfs_namenode.py

```diff
@@ -0,0 +1,189 @@
+'''
+
+# stdlib
+from urlparse import urljoin
+
+# 3rd party
+import requests
+from requests.exceptions import Timeout, HTTPError, InvalidURL, ConnectionError
+from simplejson import JSONDecodeError
+
+# Project
+from checks import AgentCheck
+
+# Service check names
+JMX_SERVICE_CHECK = 'hdfs_namenode.jmx.can_connect'
+
+# URL Paths
+JMX_PATH = 'jmx'
+
+# Namenode Bean
+HDFS_NAMENODE_BEAN_NAME = 'Hadoop:service=NameNode,name=FSNamesystemState'
+
+# Metric types
+GAUGE = 'gauge'
+
+# HDFS metrics
+HDFS_METRICS = {
+    'CapacityTotal' : ('hdfs_namenode.capacity_total',  GAUGE),
+    'CapacityUsed' : ('hdfs_namenode.capacity_used',  GAUGE),
+    'CapacityRemaining' : ('hdfs_namenode.capacity_remaining',  GAUGE),
+    'TotalLoad' : ('hdfs_namenode.total_load',  GAUGE),
+    'FsLockQueueLength' : ('hdfs_namenode.fs_lock_queue_length',  GAUGE),
+    'BlocksTotal' : ('hdfs_namenode.blocks_total',  GAUGE),
+    'MaxObjects' : ('hdfs_namenode.max_objects',  GAUGE),
+    'FilesTotal' : ('hdfs_namenode.files_total',  GAUGE),
+    'PendingReplicationBlocks' : ('hdfs_namenode.pending_replication_blocks',  GAUGE),
+    'UnderReplicatedBlocks' : ('hdfs_namenode.under_replicated_blocks',  GAUGE),
+    'ScheduledReplicationBlocks' : ('hdfs_namenode.scheduled_replication_blocks',  GAUGE),
+    'PendingDeletionBlocks' : ('hdfs_namenode.pending_deletion_blocks',  GAUGE),
+    'NumLiveDataNodes' : ('hdfs_namenode.num_live_data_nodes',  GAUGE),
+    'NumDeadDataNodes' : ('hdfs_namenode.num_dead_data_nodes',  GAUGE),
+    'NumDecomLiveDataNodes' : ('hdfs_namenode.num_decom_live_data_nodes',  GAUGE),
+    'NumDecomDeadDataNodes' : ('hdfs_namenode.num_decom_dead_data_nodes',  GAUGE),
+    'VolumeFailuresTotal' : ('hdfs_namenode.volume_failures_total',  GAUGE),
+    'EstimatedCapacityLostTotal' : ('hdfs_namenode.estimated_capacity_lost_total',  GAUGE),
+    'NumDecommissioningDataNodes' : ('hdfs_namenode.num_decommissioning_data_nodes',  GAUGE),
+    'NumStaleDataNodes' : ('hdfs_namenode.num_stale_data_nodes',  GAUGE),
+    'NumStaleStorages' : ('hdfs_namenode.num_stale_storages',  GAUGE),
+}
+
+class HDFSNameNode(AgentCheck):
+
+    def check(self, instance):
+        jmx_address = instance.get('hdfs_namenode_jmx_uri')
+        if jmx_address is None:
+            raise Exception('The JMX URL must be specified in the instance configuration')
+
+        # Get metrics from JMX
+        self._hdfs_namenode_metrics(jmx_address)
+
+    def _hdfs_namenode_metrics(self, jmx_uri):
+        '''
+        Get HDFS data node metrics from JMX
+        '''
+        response = self._rest_request_to_json(jmx_uri,
+            JMX_PATH,
+            query_params={'qry':'Hadoop:service=NameNode,name=FSNamesystemState'})
+
+        beans = response.get('beans', [])
+
+        tags = ['namenode_url:' + jmx_uri]
+
+        for bean in beans:
+            bean_name = bean.get('name')
+
+            if bean_name == HDFS_NAMENODE_BEAN_NAME:
```

Comment:
Thanks for the clarification. The only thing that bothers me a bit here is that in cases when the API doesn't return what we expect (ie no beans or multiple beans with that name), the check would either silently send no metrics or silently overwrite metrics.

Could we maybe replace the `for` loop with the following:
```python
bean = next(iter(beans))  # get the bean
bean_name = bean.get('name')
if bean_name != HDFS_NAMENODE_BEAN_NAME:
  raise Exception("Unexpected bean name {0}".format(bean_name))
```

and then loop over the metrics we want to send.

This would make the check handle unexpected results from the API more explicitly, and also make it clear in the code that we're expecting one bean only.

What do you think?